File: python/tests/test_semaphore.py
"""Tests for Semaphore implementation."""

import os
import time
import threading
import pytest
import numpy as np

from zeroipc import Memory, Semaphore


class TestSemaphoreBasic:
    """Basic Semaphore functionality tests."""

    def test_create_semaphore(self):
        """Test creating a new semaphore."""
        shm_name = f"/test_sem_create_{os.getpid()}"

        try:
            memory = Memory(shm_name, size=1024*1024)
            sem = Semaphore(memory, "test_sem", initial_count=5)

            assert sem.count == 5
            assert sem.waiting == 0
            assert sem.max_count == 0  # unbounded
            assert sem.name == "test_sem"

        finally:
            Memory.unlink(shm_name)

    def test_create_binary_semaphore(self):
        """Test creating a binary semaphore (mutex)."""
        shm_name = f"/test_sem_binary_{os.getpid()}"

        try:
            memory = Memory(shm_name, size=1024*1024)
            sem = Semaphore(memory, "mutex", initial_count=1, max_count=1)

            assert sem.count == 1
            assert sem.max_count == 1

        finally:
            Memory.unlink(shm_name)

    def test_open_existing_semaphore(self):
        """Test opening an existing semaphore."""
        shm_name = f"/test_sem_open_{os.getpid()}"

        try:
            memory = Memory(shm_name, size=1024*1024)

            # Create semaphore
            sem1 = Semaphore(memory, "existing", initial_count=3, max_count=10)
            assert sem1.count == 3

            # Open existing semaphore
            sem2 = Semaphore(memory, "existing")
            assert sem2.count == 3
            assert sem2.max_count == 10

        finally:
            Memory.unlink(shm_name)

    def test_acquire_release(self):
        """Test basic acquire and release operations."""
        shm_name = f"/test_sem_acq_rel_{os.getpid()}"

        try:
            memory = Memory(shm_name, size=1024*1024)
            sem = Semaphore(memory, "test", initial_count=3)

            # Acquire decrements count
            assert sem.acquire() == True
            assert sem.count == 2

            assert sem.acquire() == True
            assert sem.count == 1

            # Release increments count
            sem.release()
            assert sem.count == 2

            sem.release()
            assert sem.count == 3

        finally:
            Memory.unlink(shm_name)

    def test_try_acquire_success(self):
        """Test try_acquire when permits are available."""
        shm_name = f"/test_sem_try_success_{os.getpid()}"

        try:
            memory = Memory(shm_name, size=1024*1024)
            sem = Semaphore(memory, "test", initial_count=2)

            assert sem.try_acquire() == True
            assert sem.count == 1

            assert sem.try_acquire() == True
            assert sem.count == 0

        finally:
            Memory.unlink(shm_name)

    def test_try_acquire_failure(self):
        """Test try_acquire when no permits available."""
        shm_name = f"/test_sem_try_fail_{os.getpid()}"

        try:
            memory = Memory(shm_name, size=1024*1024)
            sem = Semaphore(memory, "test", initial_count=0)

            assert sem.try_acquire() == False
            assert sem.count == 0

        finally:
            Memory.unlink(shm_name)

    def test_acquire_timeout(self):
        """Test acquire with timeout when no permits available."""
        shm_name = f"/test_sem_timeout_{os.getpid()}"

        try:
            memory = Memory(shm_name, size=1024*1024)
            sem = Semaphore(memory, "test", initial_count=0)

            # Should timeout
            start = time.time()
            result = sem.acquire(timeout=0.1)
            elapsed = time.time() - start

            assert result == False
            assert elapsed >= 0.1

        finally:
            Memory.unlink(shm_name)

    def test_acquire_timeout_success(self):
        """Test acquire with timeout when permit is available."""
        shm_name = f"/test_sem_timeout_success_{os.getpid()}"

        try:
            memory = Memory(shm_name, size=1024*1024)
            sem = Semaphore(memory, "test", initial_count=1)

            # Should succeed immediately
            result = sem.acquire(timeout=0.1)
            assert result == True
            assert sem.count == 0

        finally:
            Memory.unlink(shm_name)

    def test_max_count_enforced(self):
        """Test that max_count is enforced on release."""
        shm_name = f"/test_sem_max_{os.getpid()}"

        try:
            memory = Memory(shm_name, size=1024*1024)
            sem = Semaphore(memory, "test", initial_count=5, max_count=5)

            # Try to exceed max count
            with pytest.raises(OverflowError):
                sem.release()

            assert sem.count == 5

        finally:
            Memory.unlink(shm_name)

    def test_invalid_arguments(self):
        """Test invalid constructor arguments."""
        shm_name = f"/test_sem_invalid_{os.getpid()}"

        try:
            memory = Memory(shm_name, size=1024*1024)

            # Negative initial count
            with pytest.raises(ValueError):
                Semaphore(memory, "test", initial_count=-1)

            # Negative max count
            with pytest.raises(ValueError):
                Semaphore(memory, "test", initial_count=5, max_count=-1)

            # Initial > max
            with pytest.raises(ValueError):
                Semaphore(memory, "test", initial_count=10, max_count=5)

        finally:
            Memory.unlink(shm_name)


class TestSemaphoreConcurrency:
    """Concurrency tests for Semaphore."""

    def test_mutual_exclusion(self):
        """Test using semaphore as a mutex."""
        shm_name = f"/test_sem_mutex_{os.getpid()}"

        try:
            memory = Memory(shm_name, size=1024*1024)
            sem = Semaphore(memory, "mutex", initial_count=1, max_count=1)

            shared_counter = [0]  # Use list for mutability
            iterations = 50

            def worker():
                for _ in range(iterations):
                    sem.acquire()
                    # Critical section
                    temp = shared_counter[0]
                    time.sleep(0.0001)  # Force context switch
                    shared_counter[0] = temp + 1
                    sem.release()

            threads = []
            for _ in range(4):
                t = threading.Thread(target=worker)
                t.start()
                threads.append(t)

            for t in threads:
                t.join()

            # All increments should be counted
            assert shared_counter[0] == iterations * 4

        finally:
            Memory.unlink(shm_name)

    def test_resource_pool_limiting(self):
        """Test semaphore limits concurrent access."""
        shm_name = f"/test_sem_pool_{os.getpid()}"

        try:
            memory = Memory(shm_name, size=1024*1024)
            sem = Semaphore(memory, "pool", initial_count=3)  # max 3 concurrent

            concurrent_users = [0]
            max_concurrent = [0]
            lock = threading.Lock()

            def worker():
                sem.acquire()

                # Inside critical region
                with lock:
                    concurrent_users[0] += 1
                    if concurrent_users[0] > max_concurrent[0]:
                        max_concurrent[0] = concurrent_users[0]

                time.sleep(0.01)

                with lock:
                    concurrent_users[0] -= 1

                sem.release()

            threads = []
            for _ in range(10):
                t = threading.Thread(target=worker)
                t.start()
                threads.append(t)

            for t in threads:
                t.join()

            # Never exceeded the limit
            assert max_concurrent[0] <= 3

        finally:
            Memory.unlink(shm_name)

    def test_context_manager(self):
        """Test semaphore as context manager."""
        shm_name = f"/test_sem_context_{os.getpid()}"

        try:
            memory = Memory(shm_name, size=1024*1024)
            sem = Semaphore(memory, "mutex", initial_count=1, max_count=1)

            with sem:
                assert sem.count == 0  # Acquired

            assert sem.count == 1  # Released

        finally:
            Memory.unlink(shm_name)

    def test_context_manager_exception(self):
        """Test context manager releases on exception."""
        shm_name = f"/test_sem_ctx_exc_{os.getpid()}"

        try:
            memory = Memory(shm_name, size=1024*1024)
            sem = Semaphore(memory, "mutex", initial_count=1, max_count=1)

            try:
                with sem:
                    assert sem.count == 0
                    raise RuntimeError("test exception")
            except RuntimeError:
                pass

            # Should be released despite exception
            assert sem.count == 1

        finally:
            Memory.unlink(shm_name)

    def test_multiple_waiters(self):
        """Test multiple threads waiting on semaphore."""
        shm_name = f"/test_sem_waiters_{os.getpid()}"

        try:
            memory = Memory(shm_name, size=1024*1024)
            sem = Semaphore(memory, "test", initial_count=0)

            acquired_count = [0]
            lock = threading.Lock()

            def waiter():
                sem.acquire()
                with lock:
                    acquired_count[0] += 1

            # Start 5 waiting threads
            threads = []
            for _ in range(5):
                t = threading.Thread(target=waiter)
                t.start()
                threads.append(t)

            # Give threads time to start waiting
            time.sleep(0.1)
            assert sem.waiting == 5

            # Release permits one by one
            for _ in range(5):
                sem.release()
                time.sleep(0.05)  # Let waiter wake up

            for t in threads:
                t.join()

            assert acquired_count[0] == 5
            assert sem.waiting == 0

        finally:
            Memory.unlink(shm_name)

    def test_unbounded_semaphore(self):
        """Test unbounded semaphore (max_count=0)."""
        shm_name = f"/test_sem_unbounded_{os.getpid()}"

        try:
            memory = Memory(shm_name, size=1024*1024)
            sem = Semaphore(memory, "unbounded", initial_count=0, max_count=0)

            # Can release many times
            for _ in range(100):
                sem.release()

            assert sem.count == 100

            # Can acquire all
            for _ in range(100):
                assert sem.try_acquire() == True

            assert sem.count == 0

        finally:
            Memory.unlink(shm_name)

    def test_repr(self):
        """Test string representation."""
        shm_name = f"/test_sem_repr_{os.getpid()}"

        try:
            memory = Memory(shm_name, size=1024*1024)
            sem = Semaphore(memory, "test", initial_count=5, max_count=10)

            repr_str = repr(sem)
            assert "test" in repr_str
            assert "5" in repr_str
            assert "10" in repr_str

        finally:
            Memory.unlink(shm_name)

File: cpp/tests/test_semaphore.cpp
#include <gtest/gtest.h>
#include <zeroipc/memory.h>
#include <zeroipc/semaphore.h>
#include <thread>
#include <vector>
#include <chrono>
#include <unistd.h>
#include "test_config.h"

using namespace zeroipc;
using namespace std::chrono_literals;
using namespace zeroipc::test;

class SemaphoreTest : public ::testing::Test {
protected:
    void SetUp() override {
        shm_name_ = "/test_semaphore_" + std::to_string(getpid());
    }

    void TearDown() override {
        Memory::unlink(shm_name_);
    }

    std::string shm_name_;
};

// Basic tests

TEST_F(SemaphoreTest, CreateSemaphore) {
    Memory mem(shm_name_, 1024 * 1024);
    Semaphore sem(mem, "test_sem", 5);  // initial count = 5

    EXPECT_EQ(sem.count(), 5);
    EXPECT_EQ(sem.waiting(), 0);
    EXPECT_EQ(sem.max_count(), 0);  // unbounded
    EXPECT_EQ(sem.name(), "test_sem");
}

TEST_F(SemaphoreTest, CreateBinarySemaphore) {
    Memory mem(shm_name_, 1024 * 1024);
    Semaphore sem(mem, "mutex", 1, 1);  // binary semaphore (mutex)

    EXPECT_EQ(sem.count(), 1);
    EXPECT_EQ(sem.max_count(), 1);
}

TEST_F(SemaphoreTest, OpenExistingSemaphore) {
    Memory mem(shm_name_, 1024 * 1024);

    // Create semaphore
    {
        Semaphore sem(mem, "existing", 3, 10);
        EXPECT_EQ(sem.count(), 3);
    }

    // Open existing semaphore
    Semaphore sem2(mem, "existing");
    EXPECT_EQ(sem2.count(), 3);
    EXPECT_EQ(sem2.max_count(), 10);
}

TEST_F(SemaphoreTest, AcquireRelease) {
    Memory mem(shm_name_, 1024 * 1024);
    Semaphore sem(mem, "test", 3);

    // Acquire decrements count
    sem.acquire();
    EXPECT_EQ(sem.count(), 2);

    sem.acquire();
    EXPECT_EQ(sem.count(), 1);

    // Release increments count
    sem.release();
    EXPECT_EQ(sem.count(), 2);

    sem.release();
    EXPECT_EQ(sem.count(), 3);
}

TEST_F(SemaphoreTest, TryAcquireSuccess) {
    Memory mem(shm_name_, 1024 * 1024);
    Semaphore sem(mem, "test", 2);

    EXPECT_TRUE(sem.try_acquire());
    EXPECT_EQ(sem.count(), 1);

    EXPECT_TRUE(sem.try_acquire());
    EXPECT_EQ(sem.count(), 0);
}

TEST_F(SemaphoreTest, TryAcquireFailure) {
    Memory mem(shm_name_, 1024 * 1024);
    Semaphore sem(mem, "test", 0);  // count = 0

    EXPECT_FALSE(sem.try_acquire());
    EXPECT_EQ(sem.count(), 0);
}

TEST_F(SemaphoreTest, AcquireTimeout) {
    Memory mem(shm_name_, 1024 * 1024);
    Semaphore sem(mem, "test", 0);

    // Should timeout
    auto start = std::chrono::steady_clock::now();
    bool result = sem.acquire_for(TestTiming::MEDIUM_TIMEOUT);
    auto elapsed = std::chrono::steady_clock::now() - start;

    EXPECT_FALSE(result);
    EXPECT_GE(elapsed, TestTiming::MEDIUM_TIMEOUT);
}

TEST_F(SemaphoreTest, AcquireTimeoutSuccess) {
    Memory mem(shm_name_, 1024 * 1024);
    Semaphore sem(mem, "test", 1);

    // Should succeed immediately
    bool result = sem.acquire_for(TestTiming::MEDIUM_TIMEOUT);
    EXPECT_TRUE(result);
    EXPECT_EQ(sem.count(), 0);
}

TEST_F(SemaphoreTest, MaxCountEnforced) {
    Memory mem(shm_name_, 1024 * 1024);
    Semaphore sem(mem, "test", 5, 5);  // max = 5

    // Try to exceed max count
    EXPECT_THROW(sem.release(), std::overflow_error);
    EXPECT_EQ(sem.count(), 5);
}

TEST_F(SemaphoreTest, InvalidArguments) {
    Memory mem(shm_name_, 1024 * 1024);

    // Negative initial count
    EXPECT_THROW(Semaphore(mem, "test", -1), std::invalid_argument);

    // Negative max count
    EXPECT_THROW(Semaphore(mem, "test", 5, -1), std::invalid_argument);

    // Initial > max
    EXPECT_THROW(Semaphore(mem, "test", 10, 5), std::invalid_argument);
}

// Thread concurrency tests

TEST_F(SemaphoreTest, MutualExclusion) {
    Memory mem(shm_name_, 1024 * 1024);
    Semaphore sem(mem, "mutex", 1, 1);  // binary semaphore

    int shared_counter = 0;
    const int iterations = 100;

    auto worker = [&]() {
        for (int i = 0; i < iterations; i++) {
            sem.acquire();
            // Critical section
            int temp = shared_counter;
            std::this_thread::sleep_for(TestTiming::CRITICAL_SECTION_DELAY);  // Force context switch
            shared_counter = temp + 1;
            sem.release();
        }
    };

    std::vector<std::thread> threads;
    for (int i = 0; i < 4; i++) {
        threads.emplace_back(worker);
    }

    for (auto& t : threads) {
        t.join();
    }

    // All increments should be counted
    EXPECT_EQ(shared_counter, iterations * 4);
}

TEST_F(SemaphoreTest, ResourcePoolLimiting) {
    Memory mem(shm_name_, 1024 * 1024);
    Semaphore sem(mem, "pool", 3);  // max 3 concurrent users

    std::atomic<int> concurrent_users{0};
    std::atomic<int> max_concurrent{0};

    auto worker = [&]() {
        sem.acquire();

        // Inside critical region
        int current = ++concurrent_users;

        // Update max
        int expected_max = max_concurrent.load();
        while (current > expected_max &&
               !max_concurrent.compare_exchange_weak(expected_max, current)) {
        }

        std::this_thread::sleep_for(TestTiming::THREAD_SYNC_DELAY);

        --concurrent_users;
        sem.release();
    };

    std::vector<std::thread> threads;
    for (int i = 0; i < 10; i++) {
        threads.emplace_back(worker);
    }

    for (auto& t : threads) {
        t.join();
    }

    // Never exceeded the limit
    EXPECT_LE(max_concurrent.load(), 3);
}

TEST_F(SemaphoreTest, ProducerConsumer) {
    Memory mem(shm_name_, 1024 * 1024);

    Semaphore items(mem, "items", 0);  // available items
    Semaphore slots(mem, "slots", 5);  // available slots

    std::vector<int> buffer;
    std::mutex buffer_mutex;

    auto producer = [&](int id) {
        for (int i = 0; i < 10; i++) {
            slots.acquire();  // Wait for slot
            {
                std::lock_guard<std::mutex> lock(buffer_mutex);
                buffer.push_back(id * 100 + i);
            }
            items.release();  // Signal item available
        }
    };

    auto consumer = [&]() {
        for (int i = 0; i < 20; i++) {  // Consume all 20 items (2 producers * 10)
            items.acquire();  // Wait for item
            {
                std::lock_guard<std::mutex> lock(buffer_mutex);
                EXPECT_FALSE(buffer.empty());
                buffer.pop_back();
            }
            slots.release();  // Signal slot available
        }
    };

    std::thread p1(producer, 1);
    std::thread p2(producer, 2);
    std::thread c(consumer);

    p1.join();
    p2.join();
    c.join();

    EXPECT_EQ(buffer.size(), 0);  // All items consumed
}

TEST_F(SemaphoreTest, SemaphoreGuard) {
    Memory mem(shm_name_, 1024 * 1024);
    Semaphore sem(mem, "mutex", 1, 1);

    {
        SemaphoreGuard guard(sem);
        EXPECT_EQ(sem.count(), 0);  // Acquired
        // Automatic release on scope exit
    }

    EXPECT_EQ(sem.count(), 1);  // Released
}

TEST_F(SemaphoreTest, SemaphoreGuardException) {
    Memory mem(shm_name_, 1024 * 1024);
    Semaphore sem(mem, "mutex", 1, 1);

    try {
        SemaphoreGuard guard(sem);
        EXPECT_EQ(sem.count(), 0);
        throw std::runtime_error("test exception");
    } catch (...) {
        // Guard should have released
    }

    EXPECT_EQ(sem.count(), 1);  // Released despite exception
}

TEST_F(SemaphoreTest, MultipleWaiters) {
    Memory mem(shm_name_, 1024 * 1024);
    Semaphore sem(mem, "test", 0);

    std::atomic<int> acquired{0};

    auto waiter = [&]() {
        sem.acquire();
        acquired++;
    };

    // Start 5 waiting threads
    std::vector<std::thread> threads;
    for (int i = 0; i < 5; i++) {
        threads.emplace_back(waiter);
    }

    // Give threads time to start waiting
    std::this_thread::sleep_for(TestTiming::MEDIUM_TIMEOUT);
    EXPECT_EQ(sem.waiting(), 5);

    // Release permits one by one
    for (int i = 0; i < 5; i++) {
        sem.release();
        std::this_thread::sleep_for(TestTiming::SHORT_TIMEOUT);  // Let waiter wake up
    }

    for (auto& t : threads) {
        t.join();
    }

    EXPECT_EQ(acquired.load(), 5);
    EXPECT_EQ(sem.waiting(), 0);
}

TEST_F(SemaphoreTest, UnboundedSemaphore) {
    Memory mem(shm_name_, 1024 * 1024);
    Semaphore sem(mem, "unbounded", 0, 0);  // unbounded

    // Can release many times
    for (int i = 0; i < 100; i++) {
        sem.release();
    }

    EXPECT_EQ(sem.count(), 100);

    // Can acquire all
    for (int i = 0; i < 100; i++) {
        EXPECT_TRUE(sem.try_acquire());
    }

    EXPECT_EQ(sem.count(), 0);
}

int main(int argc, char** argv) {
    ::testing::InitGoogleTest(&argc, argv);
    return RUN_ALL_TESTS();
}
